Work with application default credential

akm opened the pull request:
## Assign authorization to storage service

In order to work on local environment with [application default credential](https://developers.google.com/identity/protocols/application-default-credentials), pass `project` to `Google::Cloud::Storage.new` or `Google::Cloud::Pubsub.new`.

## Refactor module structure

- Add `Magellan::Gcs::Proxy::GCP` module to retrieve service object.
- Remove `file_operation` and `pubsub_operation`

## Fix some bugs

- Make directory before downloading
- Use logger_pipe for logging COMMAND outputs to log
- Allow message without upload_files

## Reviewers

- [x] @nagachika 
- [x] @yodack

akm committed:
:recycle: Move #storage method from FileOperation to Context

akm committed:
:shower: Remove FileOperation

akm committed:
:recycle: Extract GCP module to get services

akm committed:
:+1: Add subscription method to GCP and Use it in Cli

akm committed:
:shower: Remove pubsub_operation

akm committed:
:+1: Use application default credential

See
https://developers.google.com/identity/protocols/application-default-credentials
and https://github.com/google/google-auth-library-ruby

akm committed:
Merge remote-tracking branch 'origin/master' into features/use_google-api-client_directly

akm committed:
:bug: Bump up version from 0.1.0 to 0.1.1

akm committed:
:+1: Use application defautl credential for GCS

akm committed:
:+1: Add description about debugging to README

akm committed:
:bug: Use Log in GCP for logger

akm committed:
:+1: Make directory before download

akm committed:
:+1: Use logger_pipe to run command with logger

akm committed:
:bug: Allow message without upload_files

And add logging around each upload

nagachika commented on lib/magellan/gcs/proxy/gcp.rb:
```diff
@@ -0,0 +1,61 @@
+            #   # keyfile: ENV['GOOGLE_KEY_JSON_FILE'],
+            # )
+            credentials = Google::Cloud::Storage::Credentials.default scope: SCOPE_STORAGE
+            service = Google::Cloud::Storage::Service.new(nil, credentials, retries: nil, timeout: nil)
```
Google::Cloud::Storage.new の wrapper メソッドを使えば  credential を省略して default credential を使った Google::Cloud::Storage::Project が作れそうです。 Project には service の accessor もあるし、おそらく project のまま使っても必要な操作はできそうです。

nagachika replied:
リンク貼るのを忘れてました https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-storage/lib/google/cloud/storage.rb#L429-L437

akm replied:
@nagachika 試してみたのですが、 https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-storage/lib/google/cloud/storage.rb#L433 でエラーになってしまいます。

```
$ bundle console
[1] pry(main)> storage = Google::Cloud::Storage.new
ArgumentError: project is missing
from $HOME/.anyenv/envs/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/google-cloud-storage-0.21.0/lib/google/cloud/storage.rb:433:in `new'
```

これを回避し、かつ`Google::Apis::StorageV1::StorageService` に`authorization` を設定するために、こんな感じにしたのでした。

追記: 上記は全てローカルの環境で実行しています。

nagachika replied:
なるほどー。 環境変数 or metadata server から project-id を受けとるようにしてみてはどうでしょう。
GCE(GKE)上からは project id は `project-id` というキーで取得できます。 https://cloud.google.com/compute/docs/storing-retrieving-metadata#default (つまり `curl "http://metadata.google.internal/computeMetadata/v1/project/project-id" -H "Metadata-Flavor: Google"` 相当の HTTP リクエストで取得できます)
ローカルでは環境変数から受け取るようにすればいいと思います。

akm replied:
@nagachika それも考えたのですが、 `Google::Cloud::Storage::Service` で @project に値が設定されている必要があるのは以下の２つのメソッドだけで、これらは使っていないのでprojectにnilを設定しても現時点では問題ないと判断しました(将来的にこんな使い方は良くないだろうなとは思ってはいるのですが)。

- [#list_buckets](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-storage/lib/google/cloud/storage/service.rb#L63-L68)
- [#insert_bucket](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-storage/lib/google/cloud/storage/service.rb#L80-L87)

nagachika commented on lib/magellan/gcs/proxy/gcp.rb:
```diff
@@ -0,0 +1,61 @@
+
+        def subscription
+          unless @subscription
+            @subscription = pubsub.subscription(ENV['BLOCKS_BATCH_PUBSUB_SUBSCRIPTION'] || 'test-subscription')
```
(https://github.com/groovenauts/magellan-gcs-proxy/pull/3/files#r89036930 への返信です)
なるほど。それをきいてなんとなく pubsub のほうは topic 名や subscription 名を google-cloud-ruby 側で組み立ててるんじゃないかという予感がしたのでコードをみてみたのですが
https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/project.rb#L389-L399
から
https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/subscription.rb#L63
https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/service.rb#L344-L347
と呼ばれていてやっぱり中で組み立てていました。
従って BLOCKS_BATCH_PUBSUB_SUBSCRIPTION をそのままわたすとだめで project_id と basename の部分に分解する必要がありました。

akm replied:
@nagachika 現時点では `BLOCKS_BATCH_PUBSUB_SUBSCRIPTION ` がスラッシュを含んでいるので 
https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/service.rb#L345 で 引数`subscription_name` がそのまま返されているので動いてるのだと思います。
ですので  project_id と basename の部分に分解しなくても問題ないと思います。

akm replied:
@nagachika 色々考えて、やっぱりイレギュラーな使い方をしているので、Google::Cloud::Storage.newの引数projectに 環境変数`BLOCKS_BATCH_PROJECT_ID` あるいは  `metadata server`から `project-id` から取得した値を指定するように変更しようと思います。

`BLOCKS_BATCH_PUBSUB_SUBSCRIPTION` については分解せずにそのままでも動くと思いますがどうしましょう？

nagachika replied:
では subscription はそのままでいいと思います。project 名は deploy 側で pubsub の subscription には利用している service account の project_id をそのまま使っているので、subscription 名から分解したものを使ってもいいと思います。
または環境変数の受け取る仕様を変更して PROJECT_ID を別に受け取る(`BLOCKS_BATCH_PUBSUB_SUBSCRIPTION` などには最後の basename 部分だけ受け取る)に変更してもいいと思います(google-api-client だと自分で組み立てるのでこうしてましたが google-cloud-ruby は中で組み立てる前提ならそのほうが自然ではあるので)。
環境変数は明示的に遮蔽しなければ呼び出すコマンドにも渡されると思うので、そちらでの期待にあわせてもいいかと思います.

akm replied:
@nagachika 

> google-cloud-ruby は中で組み立てる前提ならそのほうが自然ではあるので

そうですね。では、`BLOCKS_BATCH_PUBSUB_SUBSCRIPTION` はbasename部分だけを渡してもらうように変更していただいていいですか？

akm replied:
@nagachika `project`を指定するように変更し、`authorization`に関する処理は行わないように変更しました。ローカルでの動作も確認しております。`BLOCKS_BATCH_PUBSUB_SUBSCRIPTION` は basename と 正規化された文字列 のどちらでも動きました。

akm committed:
:+1: Use project_id to build storage and pubsub

Use $BLOCKS_BATCH_PROJECT_ID or
http://metadata.google.internal/computeMetadata/v1/project/project-id
as their project.

akm committed:
:green_heart: Add spec for GCP.project_id
